src/dynamics/mpas/dyn_grid.F90

```diff
@@ -1,13 +1,457 @@
+    character(*), parameter :: dyn_grid_name(*) = [ character(max_hcoordname_len) :: &
+        'mpas_cell',  &
+        'cam_cell',   &
+        'mpas_edge',  &
+        'mpas_vertex' &
+    ]
+
+    real(kind_r8), parameter :: deg_to_rad = pi / 180.0_kind_r8 ! Convert degrees to radians.
+    real(kind_r8), parameter :: rad_to_deg = 180.0_kind_r8 / pi ! Convert radians to degrees.
+
+    ! Local and global mesh dimensions.
+    integer :: ncells_solve,  nedges_solve,  nvertices_solve
+    integer :: ncells_global, nedges_global, nvertices_global, nvertlevels, ncells_max, nedges_max
+    real(kind_r8) :: sphere_radius
 contains
+    !> Initialize various model grids (e.g., dynamics, physics) in terms of dynamics decomposition.
+    !> Additionally, MPAS framework initialization and reading time-invariant (e.g., grid/mesh) variables
+    !> are also being completed in this subroutine.
+    !> (KCW, 2024-03-29)
+    !
+    ! Called by `cam_init` in `src/control/cam_comp.F90`.
+    subroutine model_grid_init()
+        character(*), parameter :: subname = 'dyn_grid::model_grid_init'
+        integer, allocatable :: ncellssolve, nedgessolve, nverticessolve
+        type(file_desc_t), pointer :: pio_file => null()
+
+        ! Initialize mathematical and physical constants for dynamics.
+        call dyn_debug_print('Calling dynconst_init')
+
+        call dynconst_init()
+
+        ! Initialize vertical coordinates.
+        ! `pver` comes from CAM-SIMA namelist.
+        ! `pverp` is set only after this call. Call it as soon as possible.
+        call dyn_debug_print('Calling vert_coord_init')
+
+        call vert_coord_init(1, pver)
+
+        ! Get PIO file descriptor for initial file.
+        pio_file => initial_file_get_id()
+
+        ! Finish MPAS framework initialization, including
+        ! the allocation of all blocks and fields managed by MPAS.
+        call mpas_dynamical_core % init_phase3(num_advected, pio_file)
+
+        ! Read time-invariant (e.g., grid/mesh) variables.
+        call mpas_dynamical_core % read_write_stream(pio_file, 'r', 'invariant')
+
+        ! Inquire local and global mesh dimensions and save them as module variables.
+        call dyn_debug_print('Inquiring local and global mesh dimensions')
+
+        call mpas_dynamical_core % get_variable_value(ncellssolve, 'dim', 'nCellsSolve')
+        call mpas_dynamical_core % get_variable_value(nedgessolve, 'dim', 'nEdgesSolve')
+        call mpas_dynamical_core % get_variable_value(nverticessolve, 'dim', 'nVerticesSolve')
+
+        ncells_solve = ncellssolve
+        nedges_solve = nedgessolve
+        nvertices_solve = nverticessolve
+
+        deallocate(ncellssolve)
+        deallocate(nedgessolve)
+        deallocate(nverticessolve)
+
+        call mpas_dynamical_core % get_global_mesh_dimension( &
+            ncells_global, nedges_global, nvertices_global, nvertlevels, ncells_max, nedges_max, &
+            sphere_radius)
```

Comment:
`sphere_radius` is a double-precision floating point attribute **in the input file**. This attribute is MPAS-specific, and it is one of the required attributes when defining an MPAS mesh. See page 9 of "MPAS Mesh Specification" [[1]](https://mpas-dev.github.io/files/documents/MPAS-MeshSpec.pdf) for details.

The MPAS-specific input files are generated by the `init_atmosphere_model` program, which is part of MPAS. When MPAS is built as a dynamical core in CAM/CAM-SIMA (i.e., the `MPAS_CAM_DYCORE` macro is defined at compile-time), `sphere_radius` is set to `rearth` from the `physconst` module [[2]](https://github.com/ESCOMP/CAM/blob/cam_development/src/utils/physconst.F90#L92)[[3]](https://github.com/ESCOMP/CAM-SIMA/blob/development/src/data/physconst.F90#L88), which is part of CAM/CAM-SIMA, through the following chain of events:

1. In the `mpas_constants` module, `a` is set to `rearth` from the `physconst` module [[4]](https://github.com/MPAS-Dev/MPAS-Model/blob/v8.0.2/src/framework/mpas_constants.F#L29).
2. In the `init_atm_cases` module, the `mpas_constants` module above is `use`d, importing all its module variables [[5]](https://github.com/MPAS-Dev/MPAS-Model/blob/v8.0.2/src/core_init_atmosphere/mpas_init_atm_cases.F#L13).
3. In the `init_atm_core` module, the `init_atm_cases` module above is `use`d, again importing all its module variables [[6]](https://github.com/MPAS-Dev/MPAS-Model/blob/v8.0.2/src/core_init_atmosphere/mpas_init_atm_core.F#L19). Shortly after, `sphere_radius` is set to `a` [[7]](https://github.com/MPAS-Dev/MPAS-Model/blob/v8.0.2/src/core_init_atmosphere/mpas_init_atm_core.F#L49), and is written to the generated input file [[8]](https://github.com/MPAS-Dev/MPAS-Model/blob/v8.0.2/src/core_init_atmosphere/mpas_init_atm_core.F#L55).

As long as the definition of `rearth` does not change in the `physconst` module, `sphere_radius` in the generated input file should be exactly the same as `rearth`.

Reply:
Since input files are generated by the `init_atmosphere` core, which is not included in the MPAS-A dycore implementation in CAM-SIMA, I think it would in principle be possible for the `sphere_radius` in the input file to be different from what CAM-SIMA specifies in its `physconst` module. The `init_atmosphere` core uses 6371229.0 m as the earth radius: https://github.com/MPAS-Dev/MPAS-Model/blob/v8.1.0/src/framework/mpas_constants.F#L38 . (Side note, we really should rename our variable from `a` to `r_sphere` or something else in MPAS!)

Reply:
@kuanchihwang @mgduda Thanks for the explanation!  It sounds like there is theoretically a risk, but only for currently uncommon simulation setups like exoplanets, in which case the user might need to be more careful with their input files (which they likely would have to do anyways).  So I am happy to not worry about this particular problem, at least for now.